### PR TITLE
[Fix](Nereids) Add cast comparison with slot reference when inferring predicate

### PR DESCRIPTION
**Pull request by LiBinfeng-01**
## Proposed changes

Problem: 
When inferring predicate, we assume that slot reference need to be inferred. But in this case:
carete table tb1(l1 smallint) ...;
create table tb2(l2 int) ...;
select * from tb1 inner join tb2 where tb1.l1 = tb2.l2  and tb2.l2 = 1; 
We can not get tb1.l1 = 1 filter because we will add a cast to l1 (Cast smallint to int l1) = l2.

Solved:
Add cast consideration when inferring predicate, also add change judgement when judging equals to slotreference and cast expression. But when we want to infer predicate from bigger type cast to smaller type, it is logical error. 
For example: 
select * from tb1 inner join tb2 where tb1.l1 = cast(tb2.l2 as smallint)  and tb2.l2 = (number between smallint max and intmax); 
tb2.l2 value can not infer to left side because tb1.l1 would be false value, and when we add one more condition like tb1.l1 = tb3.l3(smallint). It would cause this predicate be false.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run feut

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

File: regression-test/suites/nereids_p0/infer_predicate/load.groovy [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+suite("load") {
+    sql 'create database if not exists nereids_infer_predicate_test'
```
Comment (morrySnow) [inline regression-test/suites/nereids_p0/infer_predicate/load.groovy:19]: not need create database here, remove load.groovy completely

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java:1]: shoud use getExpressionCoveredByCast to process expression like `cast(cast(cast...(a)))`

**Review by morrySnow — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java:1]: get

File: regression-test/suites/nereids_p0/infer_predicate/load.groovy (hunk 1 above)
Comment (LiBinfeng-01) [reply regression-test/suites/nereids_p0/infer_predicate/load.groovy]: done

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.08 seconds
 stream load tsv:          457 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          56 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          68.2 seconds inserted 10000000 Rows, about 146K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230705071925_clickbench_pr_172566.html

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.46 seconds
 stream load tsv:          457 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          56 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          67.7 seconds inserted 10000000 Rows, about 147K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230705142027_clickbench_pr_172936.html

**Comment by hello-stephen** [comment]
(From new mechine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 56.08 seconds
 stream load tsv:          516 seconds loaded 74807831229 Bytes, about 138 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          87.3 seconds inserted 10000000 Rows, about 114K ops/s
 storage size: 17167412237 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230705231101_clickbench_pr_172941.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 52.65 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          90.5 seconds inserted 10000000 Rows, about 110K ops/s
 storage size: 17167473914 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230706181641_clickbench_pr_173630.html

**Review by starocean999 — DISMISSED** [review]

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.7 seconds
 stream load tsv:          456 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          67.7 seconds inserted 10000000 Rows, about 147K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707011857_clickbench_pr_174031.html

**Comment by LiBinfeng-01** [comment]
run clickbench

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.53 seconds
 stream load tsv:          502 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          88.8 seconds inserted 10000000 Rows, about 112K ops/s
 storage size: 17166995123 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707114339_clickbench_pr_174139.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.61 seconds
 stream load tsv:          449 seconds loaded 74807831229 Bytes, about 158 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          68.8 seconds inserted 10000000 Rows, about 145K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707041006_clickbench_pr_174123.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.13 seconds
 stream load tsv:          447 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          68.7 seconds inserted 10000000 Rows, about 145K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707043705_clickbench_pr_174136.html

**Review by starocean999 — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java [hunk 2]
```diff
@@ -70,19 +72,44 @@ public Expression visit(Expression expr, Void context) {
...
                 }
                 return super.visit(cp, context);
             }
 
-            private Expression replaceSlot(Expression expr) {
+            private boolean isTwoExpressionEqualWithCast(Expression left, Expression right) {
+                return getSlotCoveredByCast(left).equals(getSlotCoveredByCast(right));
+            }
+
+            private boolean canCompareType(DataType dataType) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java:87]: if u want get all integer type , u could use `instance of IntegralType`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java [hunk 3]
```diff
@@ -70,19 +72,44 @@ public Expression visit(Expression expr, Void context) {
...
+                        || dataType.isTinyIntType() || dataType.isLargeIntType();
+            }
+
+            private boolean isOriginDataTypeBigger(DataType originDataType, Expression expr) {
+                if (canCompareType(leftSlotEqualToRightSlot.child(0).getDataType())
+                        && canCompareType(leftSlotEqualToRightSlot.child(1).getDataType())
+                                && canCompareType(originDataType)) {
+                    // infer filter can not be lower than original datatype, or dataset would be wrong
+                    if (originDataType.width() > leftSlotEqualToRightSlot.child(0).getDataType().width()
+                            || originDataType.width() > leftSlotEqualToRightSlot.child(1).getDataType().width()) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java:98]: use width not a good idea. maybe u should add widerThan or similar method to IntegralType

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java [hunk 4]
```diff
@@ -92,13 +119,48 @@ private Expression replaceSlot(Expression expr) {
...
+    /**
+     * get slot covered by cast
+     * example: input: cast(cast(table.columnA)) output: table.columnA
+     */
+    private Expression getSlotCoveredByCast(Expression cast) {
+        if (cast instanceof Cast) {
+            return getSlotCoveredByCast(((Cast) cast).child());
+        }
+        return cast;
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java:154]: move it into ExpressionUtils

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java [hunk 5]
```diff
@@ -92,13 +119,48 @@ private Expression replaceSlot(Expression expr) {
...
+    /**
+     * judge if expression is slot covered by cast
+     * example: cast(cast(table.columnA))
+     */
+    private boolean isExpressionSlotCoveredByCast(Expression expr) {
+        if (expr instanceof Cast) {
+            return isExpressionSlotCoveredByCast(((Cast) expr).child());
+        }
+        return expr instanceof SlotReference;
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java:131]: move it into ExpressionUtils

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java [hunk 6]
```diff
@@ -70,19 +72,44 @@ public Expression visit(Expression expr, Void context) {
...
+                } else if (isExpressionSlotCoveredByCast(cp.right()) && cp.left().isConstant()) {
+                    return replaceSlot(cp, getDatatypeCoveredByCast(cp.right()));
                 }
                 return super.visit(cp, context);
             }
 
-            private Expression replaceSlot(Expression expr) {
+            private boolean isTwoExpressionEqualWithCast(Expression left, Expression right) {
+                return getSlotCoveredByCast(left).equals(getSlotCoveredByCast(right));
+            }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java:85]: move it into ExpressionUtils

Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java (hunk 5 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java (hunk 4 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java (hunk 3 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java (hunk 2 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java]: done

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 54.66 seconds
 stream load tsv:          503 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         18 seconds loaded 2358488459 Bytes, about 124 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          28.8 seconds inserted 10000000 Rows, about 347K ops/s
 storage size: 17169748213 Bytes

**Review by morrySnow — APPROVED** [review]

**Review by starocean999 — APPROVED** [review]

**Pull request merged**